In [2]:
import numpy as np; import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error

In [3]:
import os
current_directory = os.getcwd()
print(f"Current working directory: {current_directory}")

Current working directory: /content


In [4]:
os.chdir('/content/drive/MyDrive/Colab Notebooks/AICode/AICode586/ML/Linear Regression')
current_directory = os.getcwd()
print(f"Current working directory: {current_directory}")

Current working directory: /content/drive/MyDrive/Colab Notebooks/AICode/AICode586/ML/Linear Regression


In [5]:
college = pd.read_excel('college_dataset.xlsx'); college.sample(2)

,Unnamed: 0,Private,Apps,Accept,Enroll,Top10perc,Top25perc,F.Undergrad,P.Undergrad,Outstate,Room.Board,Books,Personal,PhD,Terminal,S.F.Ratio,perc.alumni,Expend,Grad.Rate
731,Webber College,Yes,280,143,79,5,27,327,110,5590,2900,650,1952,53,63,15.1,4,4839,90
630,University of Maine at Machias,No,441,369,172,17,45,633,317,6600,3680,600,400,46,46,15.1,4,5935,64


In [6]:
college.shape

(777, 19)

In [7]:
college.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 777 entries, 0 to 776
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   777 non-null    object 
 1   Private      777 non-null    object 
 2   Apps         777 non-null    int64  
 3   Accept       777 non-null    int64  
 4   Enroll       777 non-null    int64  
 5   Top10perc    777 non-null    int64  
 6   Top25perc    777 non-null    int64  
 7   F.Undergrad  777 non-null    int64  
 8   P.Undergrad  777 non-null    int64  
 9   Outstate     777 non-null    int64  
 10  Room.Board   777 non-null    int64  
 11  Books        777 non-null    int64  
 12  Personal     777 non-null    int64  
 13  PhD          777 non-null    int64  
 14  Terminal     777 non-null    int64  
 15  S.F.Ratio    777 non-null    float64
 16  perc.alumni  777 non-null    int64  
 17  Expend       777 non-null    int64  
 18  Grad.Rate    777 non-null    int64  
dtypes: float

# Task 1
### Data Cleaning
- Check for missing values
- Remove the column containing university names
- Do not perform any additional cleaning

In [ ]:
##### CODE HERE #####

# Task 2
### Statistics Analysis
- Generate descriptive statistics for numeric columns
- Compute Pearson correlation between numerical features and identify the pair of features that are most correlated

In [ ]:
##### CODE HERE #####

# Task 3
### Target Variable
- Set `Grad.Rate` as the target variable
- All other columns are treated as predictors

In [ ]:
##### CODE HERE #####

# Task 4
### Data Preprocessing
- Do not scale or normalise the data
- Convert categorical variables into numeric format using one-hot encoding with `pd.get_dummies()` with `drop_first = True` to avoid multicollinearity

In [ ]:
##### CODE HERE #####

# Task 5
### Train-Validation Split
- Split the data into training (70%) and validation (30%) sets using `train_test_split()` with `random_state = 42` for reproducibility

In [ ]:
##### CODE HERE #####

# Task 6
### Model Training
- Fit a linear regression model on the training data using `LinearRegression()` from `sklearn`

In [ ]:
##### CODE HERE #####

# Task 7
### Model Parameters
- Display model coefficients and intercept values using the attributes `.coef_` and `.intercept_`

In [ ]:
##### CODE HERE #####

# Task 8
### Predictions
- Generate predictions on both training and testing sets

In [ ]:
##### CODE HERE #####

# Task 9
### Model Performance
- Evaluate model performance using R² and MAE on both training and testing sets

In [ ]:
##### CODE HERE #####

# Task
Build and evaluate a linear regression model to predict `Grad.Rate` using the "college_dataset.xlsx" dataset. The process involves checking for missing values, removing the 'Unnamed: 0' column, generating descriptive statistics for numeric columns, computing Pearson correlation, setting `Grad.Rate` as the target variable, one-hot encoding categorical variables with `drop_first = True`, splitting the data into 70% training and 30% validation sets with `random_state = 42`, training a `LinearRegression()` model, displaying its coefficients and intercept, generating predictions, and finally, evaluating the model's performance using R² and MAE on both training and testing sets to summarize the findings.

## Data Cleaning

### Subtask:
Check for missing values and remove the 'Unnamed: 0' column as instructed in Task 1. Do not perform any additional cleaning.


**Reasoning**:
First, I'll check for missing values to understand the data completeness. Then, I will remove the 'Unnamed: 0' column as instructed, which is likely an artifact from data loading.



In [8]:
print("Missing values before cleaning:")
print(college.isnull().sum())

# Remove the 'Unnamed: 0' column
college.drop('Unnamed: 0', axis=1, inplace=True)
print("\n'Unnamed: 0' column removed. Displaying head of the updated DataFrame:")
college.head()

Missing values before cleaning:
Unnamed: 0     0
Private        0
Apps           0
Accept         0
Enroll         0
Top10perc      0
Top25perc      0
F.Undergrad    0
P.Undergrad    0
Outstate       0
Room.Board     0
Books          0
Personal       0
PhD            0
Terminal       0
S.F.Ratio      0
perc.alumni    0
Expend         0
Grad.Rate      0
dtype: int64

'Unnamed: 0' column removed. Displaying head of the updated DataFrame:


,Private,Apps,Accept,Enroll,Top10perc,Top25perc,F.Undergrad,P.Undergrad,Outstate,Room.Board,Books,Personal,PhD,Terminal,S.F.Ratio,perc.alumni,Expend,Grad.Rate
0,Yes,1660,1232,721,23,52,2885,537,7440,3300,450,2200,70,78,18.1,12,7041,60
1,Yes,2186,1924,512,16,29,2683,1227,12280,6450,750,1500,29,30,12.2,16,10527,56
2,Yes,1428,1097,336,22,50,1036,99,11250,3750,400,1165,53,66,12.9,30,8735,54
3,Yes,417,349,137,60,89,510,63,12960,5450,450,875,92,97,7.7,37,19016,59
4,Yes,193,146,55,16,44,249,869,7560,4120,800,1500,76,72,11.9,2,10922,15


## Statistics Analysis

### Subtask:
Generate descriptive statistics for numeric columns and compute Pearson correlation between numerical features to identify the most correlated pair, as per Task 2.


**Reasoning**:
First, I'll generate descriptive statistics for all numeric columns in the `college` DataFrame to understand their distributions and central tendencies.



In [9]:
print("Descriptive statistics for numeric columns:")
print(college.describe())

Descriptive statistics for numeric columns:
               Apps        Accept       Enroll   Top10perc   Top25perc  \
count    777.000000    777.000000   777.000000  777.000000  777.000000   
mean    3001.638353   2018.804376   779.972973   27.558559   55.796654   
std     3870.201484   2451.113971   929.176190   17.640364   19.804778   
min       81.000000     72.000000    35.000000    1.000000    9.000000   
25%      776.000000    604.000000   242.000000   15.000000   41.000000   
50%     1558.000000   1110.000000   434.000000   23.000000   54.000000   
75%     3624.000000   2424.000000   902.000000   35.000000   69.000000   
max    48094.000000  26330.000000  6392.000000   96.000000  100.000000   

        F.Undergrad   P.Undergrad      Outstate   Room.Board        Books  \
count    777.000000    777.000000    777.000000   777.000000   777.000000   
mean    3699.907336    855.298584  10440.669241  4357.526384   549.380952   
std     4850.420531   1522.431887   4023.016484  1096.6964

**Reasoning**:
Now, I will compute the Pearson correlation matrix for all numeric columns in the `college` DataFrame to understand the relationships between variables, and then identify the pair of features with the highest absolute correlation coefficient.



In [10]:
correlation_matrix = college.corr(numeric_only=True)

# Identify the most correlated pair of features (excluding self-correlation)
# Melt the correlation matrix to easily find the max/min values
correlation_df = correlation_matrix.unstack().sort_values(ascending=False)

# Get the top correlated pair (excluding self-correlation)
top_correlated_pair = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].iloc[0]
most_correlated_feature_1, most_correlated_feature_2 = top_correlated_pair.name

# Get the most negatively correlated pair
negatively_correlated_pair = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].iloc[-1]
most_negatively_correlated_feature_1, most_negatively_correlated_feature_2 = negatively_correlated_pair.name

print("\nPearson Correlation Matrix:")
print(correlation_matrix)

print(f"\nMost positively correlated pair: {most_correlated_feature_1} and {most_correlated_feature_2} (Correlation: {top_correlated_pair:.2f})")
print(f"Most negatively correlated pair: {most_negatively_correlated_feature_1} and {most_negatively_correlated_feature_2} (Correlation: {negatively_correlated_pair:.2f})")

AttributeError: 'numpy.float64' object has no attribute 'name'

**Reasoning**:
The previous code failed because it attempted to access a `.name` attribute on a float value, not the MultiIndex. I need to correctly extract the feature names from the MultiIndex of the `correlation_df` Series when identifying the most and least correlated pairs.



In [11]:
correlation_matrix = college.corr(numeric_only=True)

# Identify the most correlated pair of features (excluding self-correlation)
correlation_df = correlation_matrix.unstack().sort_values(ascending=False)

# Get the index (feature names) of the top correlated pair (excluding self-correlation)
top_correlated_value = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].iloc[0]
most_correlated_feature_1, most_correlated_feature_2 = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].index[0]

# Get the index (feature names) of the most negatively correlated pair
negatively_correlated_value = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].iloc[-1]
most_negatively_correlated_feature_1, most_negatively_correlated_feature_2 = correlation_df[correlation_df.index.get_level_values(0) != correlation_df.index.get_level_values(1)].index[-1]

print("\nPearson Correlation Matrix:")
print(correlation_matrix)

print(f"\nMost positively correlated pair: {most_correlated_feature_1} and {most_correlated_feature_2} (Correlation: {top_correlated_value:.2f})")
print(f"Most negatively correlated pair: {most_negatively_correlated_feature_1} and {most_negatively_correlated_feature_2} (Correlation: {negatively_correlated_value:.2f})")


Pearson Correlation Matrix:
                 Apps    Accept    Enroll  Top10perc  Top25perc  F.Undergrad  \
Apps         1.000000  0.943451  0.846822   0.338834   0.351640     0.814491   
Accept       0.943451  1.000000  0.911637   0.192447   0.247476     0.874223   
Enroll       0.846822  0.911637  1.000000   0.181294   0.226745     0.964640   
Top10perc    0.338834  0.192447  0.181294   1.000000   0.891995     0.141289   
Top25perc    0.351640  0.247476  0.226745   0.891995   1.000000     0.199445   
F.Undergrad  0.814491  0.874223  0.964640   0.141289   0.199445     1.000000   
P.Undergrad  0.398264  0.441271  0.513069  -0.105356  -0.053577     0.570512   
Outstate     0.050159 -0.025755 -0.155477   0.562331   0.489394    -0.215742   
Room.Board   0.164939  0.090899 -0.040232   0.371480   0.331490    -0.068890   
Books        0.132559  0.113525  0.112711   0.118858   0.115527     0.115550   
Personal     0.178731  0.200989  0.280929  -0.093316  -0.080810     0.317200   
PhD        

## Target Variable Definition

### Subtask:
Set 'Grad.Rate' as the target variable and all other columns as predictors, as specified in Task 3.


**Reasoning**:
As instructed, I will create the target variable `y` by selecting the 'Grad.Rate' column and the predictor variables `X` by selecting all other columns from the `college` DataFrame.



In [12]:
y = college['Grad.Rate']
X = college.drop('Grad.Rate', axis=1)

## Data Preprocessing

### Subtask:
Convert categorical variables into numeric format using one-hot encoding with `pd.get_dummies()` and `drop_first = True` to avoid multicollinearity, as per Task 4.


**Reasoning**:
First, I'll identify the categorical columns in the `X` DataFrame using `select_dtypes(include='object')`. Then, I will apply one-hot encoding to these columns using `pd.get_dummies()` with `drop_first=True` to convert them into a numeric format while avoiding multicollinearity, as instructed.



In [13]:
categorical_cols = X.select_dtypes(include='object').columns
X = pd.get_dummies(X, columns=categorical_cols, drop_first=True)

print("DataFrame X after one-hot encoding:")
print(X.head())

DataFrame X after one-hot encoding:
   Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   
3   417     349     137         60         89          510           63   
4   193     146      55         16         44          249          869   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   
3     12960        5450    450       875   92        97        7.7   
4      7560        4120    800      1500   76        72       11.9   

   perc.alumni  Expend  Private_Yes  
0           12    7041         True  
1           16  

## Train-Validation Split

### Subtask:
Split the data into training (70%) and validation (30%) sets using `train_test_split()` with `random_state = 42` for reproducibility, as per Task 5.


**Reasoning**:
I will split the feature and target data into training and validation sets using `train_test_split` with the specified parameters (`test_size=0.3`, `random_state=42`) and assign them to the correct variables.



In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (543, 17)
Shape of X_test: (234, 17)
Shape of y_train: (543,)
Shape of y_test: (234,)


# Task 6
### Model Training
- Fit a linear regression model on the training data using `LinearRegression()` from `sklearn`

# Task 6
### Model Training
- Fit a linear regression model on the training data using `LinearRegression()` from `sklearn`

**Reasoning**:
The previous step introduced the 'Model Training' task. Now, I need to implement the model training by initializing and fitting a `LinearRegression` model using the `X_train` and `y_train` data.



In [15]:
model = LinearRegression()
model.fit(X_train, y_train)

print("Linear Regression model fitted successfully.")

Linear Regression model fitted successfully.


# Task 7
### Model Parameters
- Display model coefficients and intercept values using the attributes `.coef_` and `.intercept_`

# Task 7
### Model Parameters
- Display model coefficients and intercept values using the attributes `.coef_` and `.intercept_`

**Reasoning**:
Now that Task 7 has been introduced, I will generate the Python code to display the trained model's coefficients and intercept using `model.coef_` and `model.intercept_` as instructed.



In [16]:
print("Model Coefficients:", model.coef_)
print("Model Intercept:", model.intercept_)

Model Coefficients: [ 1.52372919e-03 -1.37974487e-03  2.04218073e-03  3.35374363e-02
  1.44385338e-01 -1.83092082e-04 -2.00268302e-03  7.04378163e-04
  2.15058173e-03 -1.92139009e-03 -1.83456063e-03  1.41736811e-01
 -1.11153152e-01 -2.82926503e-02  3.37610792e-01 -5.32089945e-04
  5.05056314e+00]
Model Intercept: 34.93871250622216


# Task 8
### Predictions
- Generate predictions on both training and testing sets

# Task 8
### Predictions
- Generate predictions on both training and testing sets

**Reasoning**:
Now that Task 8 has been introduced, I will generate predictions for both the training and testing datasets using the trained linear regression model's `predict()` method.



In [17]:
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

print("Predictions generated for both training and testing sets.")

Predictions generated for both training and testing sets.


# Task 9
### Model Performance
- Evaluate model performance using R² and MAE on both training and testing sets

**Reasoning**:
Now that Task 9 has been introduced, I will calculate the R² and Mean Absolute Error (MAE) for both the training and testing sets using the generated predictions and actual target values, as instructed.



In [18]:
r2_train = r2_score(y_train, y_train_pred)
mae_train = mean_absolute_error(y_train, y_train_pred)

r2_test = r2_score(y_test, y_test_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

print(f"Training R\u00b2 Score: {r2_train:.2f}")
print(f"Training MAE: {mae_train:.2f}")
print(f"Testing R\u00b2 Score: {r2_test:.2f}")
print(f"Testing MAE: {mae_test:.2f}")

Training R² Score: 0.46
Training MAE: 9.37
Testing R² Score: 0.44
Testing MAE: 9.60


## Final Task

### Subtask:
Summarize the findings from the model performance evaluation and conclude the analysis.


## Summary:

### Q&A
The linear regression model built to predict `Grad.Rate` achieved an R² score of 0.46 and a Mean Absolute Error (MAE) of 9.37 on the training set. On the testing set, the model performed similarly with an R² score of 0.44 and an MAE of 9.60.

### Data Analysis Key Findings
*   The dataset initially contained no missing values and the 'Unnamed: 0' column was successfully removed.
*   Descriptive statistics provided insights into the numerical features, for example, `Grad.Rate` ranges from 10 to 118 with a mean of approximately 65.46.
*   **Correlation Analysis**:
    *   The most positively correlated pair of features was `F.Undergrad` (Full-time Undergraduates) and `Enroll` (Enrolled Students) with a Pearson correlation coefficient of 0.96.
    *   The most negatively correlated pair was `S.F.Ratio` (Student/Faculty Ratio) and `Expend` (Instructional Expenditure per Student) with a coefficient of -0.58.
*   The target variable `Grad.Rate` was separated, and all other columns became predictor variables.
*   Categorical variables, specifically the 'Private' column, were successfully one-hot encoded into `Private_Yes` using `drop_first=True` to prevent multicollinearity.
*   The data was split into training (70%, 543 samples) and testing (30%, 234 samples) sets with `random_state=42`.
*   A Linear Regression model was successfully trained on the data, yielding an intercept of 34.94.
*   **Model Performance**:
    *   **Training Set**: R² score of 0.46 and MAE of 9.37.
    *   **Testing Set**: R² score of 0.44 and MAE of 9.60.

### Insights or Next Steps
*   The model explains about 44-46% of the variance in `Grad.Rate`, indicating that while it captures some of the relationships, there is still a significant portion of variability unexplained. The similar performance on training and testing sets suggests the model is not significantly overfitting.
*   To improve model performance, consider exploring more advanced feature engineering techniques, such as creating interaction terms or polynomial features, and evaluating alternative regression models like Ridge, Lasso, or ensemble methods which might capture more complex relationships in the data.
